### PR TITLE
Improve asset resolution in Astro

### PR DESCRIPTION
**Pull request by drwpow**
## Changes

Partially addresses #96. Imagine a user has the following:

```html
<!-- src/pages/index.astro -->
<img src="./img.png" />
```

What should the final URL be? Should this be the file in `public/img.png`? Or the file in `src/pages/img.png`? Though this question is ambiguous, one thing that’s not is that this currently would break Astro trying to resolve.

This PR fixes the crash where Astro couldn’t finish a build if assets were referenced outside of `public/`. But it stops short of transforming HTML. So in this scenario, that `` would exist in the final HTML exactly as the user wrote it. For now, they would then have to decide either to replace that with:

- `"/img.png"`: I want the file in `public/`
- `"/_astro/src/pages/img.png"`: I want the file in `src/pages`

Docs were added describing this scenario

## Testing

Tests added

## Docs

Docs updated

### CODE REVIEW

File: packages/astro/src/search.ts
Comment (drwpow) [inline packages/astro/src/search.ts:131]: TL;DR there was still quite a bit of hardcoded assumptions in Astro about how to resolve URLs. While there are still a few lingering, this removes most of the common ones users may hit and get unexpected “not found” bugs

File: packages/astro/test/astro-resolve.test.js
Comment (drwpow) [inline packages/astro/test/astro-resolve.test.js:16]: Test added

File: docs/syntax.md
Comment (FredKSchott) [inline docs/syntax.md:139]: can we be more specific? "URLs and references inside of HTML element attributes"? Snowpack does resolve import URLs, for example.

File: docs/syntax.md
Comment (FredKSchott) [inline docs/syntax.md:141]: (I have my "concise docs" / "readers aren't reading all that" hat on from the other PRs) Can we make these docs more concise, and communicate the point faster? Here would be an example attempt:

---

### URL resolution

It’s important to note that Astro **won’t** transform your URL references inside of HTML for you. For example:

```html
<!-- src/components/button.astro -->
<!-- INCORRECT: This will render this exact HTML to every page its used on, 
        no matter what the URL is. Relative URLs will almost always break on some pages.  -->
<a href="./foo.css">
```

Instead, you should choose one of two options:

#### Option 1: Absolute URL References

Use absolute URLs to files in the `public/` directory whenever possible to get around this limitation. Absolute URLs will always point to the same file, regardless of what page they are rendered on.

```html
<!-- src/components/button.astro -->
<a href="/css/foo.css">
```

#### Option 2: Asset Import References

Most assets should live in the `public/` directory so that they are available via absolute URL. If you absolutely need to reference an asset inside of the `src/` directory, you can import a reference via JavaScript ESM Import:


```svelte
---
// src/components/button.astro 
// NOTE: This will only work for images, videos, and non-code assets.
import imageUrl from './cat.png';
---
<!-- src/components/button.astro -->
<img src={imageUrl}>
```

File: packages/astro/src/build/util.ts
Comment (FredKSchott) [inline packages/astro/src/build/util.ts:26]: This is important: I considered it sacred that `src/pages` only contained Markdown and Astro files. non-pages shouldn't live in `src/pages`, right? `public/` is a much better place (or `src/assets` if for some reason they need to be inside of src).

Can we instead throw a warning/error if the extension doesn't match?

**Review by FredKSchott — DISMISSED** [review]

File: packages/astro/src/build/util.ts
Comment (drwpow) [reply packages/astro/src/build/util.ts:26]: This hasn’t been the case for a while, since we decided that there is no “Astro” folder; only that some folder (which the user can configure; `src/pages` by default) maps 1:1 to the routing structure. But since day 1 we’ve said that you can import a file from anywhere in your project. That means if you imported from `src/components/*`, **that has to have some final URL.**

That last part is important: it means that we really can’t restrict users from putting additional files within `src/pages`. Not unless we did more work to restrict that, and throw errors.

This PR doesn’t change fundamentally how Astro works at all; this is a behind-the-scenes change that only fixes bugs that existed before. This PR stops short of any opinions that affect Astro’s design. If we wanted to throw an error if a user put anything in `src/pages` other than `.astro` or `.md`, that would be a new change we’d have to discuss. And that would have to happen after this PR merged anyway since it’s additional work. This is a bugfix and nothing more.

File: docs/syntax.md
Comment (drwpow) [reply docs/syntax.md:141]: I’m confused when you ask to “make these docs more concise” but you increase the word count from what I’ve written by 10% (with the same number of code blocks). Did you mean to say “add more explanation and add more headings to make it more referenceable?”

File: docs/syntax.md
Comment (drwpow) [reply docs/syntax.md:139]: Good point; this is ambiguous

File: docs/syntax.md
Comment (FredKSchott) [reply docs/syntax.md:141]: I guess I got a bit carried away there in my example :) To be fair I did add some new docs as well (Option 2)

Maybe there's a better word/phrase for what i'm trying to communicate with these docs changes. "to the point" or maybe even "easy to jump into"?

If it helps, the way that I think about it is "It should be easy to do a control-F find on the page, and then understand the example in front of them" Fun-to-read prose, diff snippets that build on previous examples, setting up a file system at the top of the page and then assuming the reader has seen it, etc. are getting the scalpel.

Comment (FredKSchott) [reply docs/syntax.md:141]: Feel free to make edits on top of mine, especially if you think it could be improved! The structure of the headings, code snippets, and overall layout of information is really the suggestion I was trying to make.

File: packages/astro/src/build/util.ts
Comment (FredKSchott) [reply packages/astro/src/build/util.ts:26]: Ah, so If you build a site now and it has an image in `src/pages`, does the site build? Dev works as well?

If so, then I'm on board with treating this comment as unrelated / out of scope for this PR. Happy to spin off in a separate issue. If not lets hold off on merging a day or two so that we don't thrash the users on this.

**Review by FredKSchott — COMMENTED** [review]
talked in person, no longer blocking (unless `src/pages/foo.png` currently breaks astro today)

File: docs/syntax.md
Comment (drwpow) [reply docs/syntax.md:141]: OK that makes sense—thanks for clarifying!

File: packages/astro/src/build/util.ts
Comment (drwpow) [reply packages/astro/src/build/util.ts:26]: > Ah, so If you build a site now and it has an image in `src/pages`, does the site build? Dev works as well?

The problem wasn’t as apparent in dev as you had to try to load that file manually. But in build, we’d scan for files to make sure they all built, and that was where the error was. But yes, overall the TL;DR of this PR is “this fixes unexpected build failures in Astro just based off HTML you may have written”

File: docs/syntax.md
Comment (drwpow) [reply docs/syntax.md]: As an extra note, it is hard to receive feedback where things are completely rewritten, including bullet points, tone, language, organization, etc to the point the original writing is completely gone. For me, at least, it’s easier to understand docs feedback where you only make specific edits to the text while keeping the same structure and tone.

Or if you have more sweeping structural feedback, a general “break this up into sections” comment is more helpful than receiving a full rewrite in the form of several paragraphs.

**Pull request merged**
